src/components/m-datepicker/index.js

```diff
@@ -2,106 +2,177 @@ import PropTypes from '../../js/prop-types'; // eslint-disable-next-line import/
-    // this.enableContext()
-
-    // or do you want to consume a specific context
-    // this.selectContext('axa-context-provider');
-  }
-
-  /**
-   * REF: https://www.w3.org/TR/custom-elements/#custom-element-conformance
-   */
-  connectedCallback() {
-    super.connectedCallback();
-    this.className = `${this.initialClassName} m-datepicker`;
-    // Your DOM interaction here, but keep it decoupled.
-    // If you don't have any, just remove this function
-  }
-
-  // You have some special logic? Or need to update the web-components DOM node itself?
-  // Then don't forget to make sure that incremental rendering works properly.
-  // attributeChangedCallback(name, oldValue, newValue) {
-  //   super.attributeChangedCallback(name, oldValue, newValue);
-  // }
-
-  // You may want to update stuff before rendering.
-  // willRenderCallback(initial) {
-  // }
-
-  // You may want to update staff after rendering
-  // didRenderCallback(initial) {
-  // }
-
-  disconnectedCallback() {
-    super.disconnectedCallback();
-
-    this.datepicker.destroy();
-  }
-  // Do you consume context?
-  // contextCallback(contextNode) {
-  //   contextNode is now available.
-  // }
-  didRenderCallback() {
-    // TODO see if it has to be today and if so pass here value
-    this.datepicker.init(getAttribute(this, 'output-iso'));
-  }
+class AXAMDatepicker extends BaseComponentGlobal {
+    static tagName = 'axa-m-datepicker'
+    static propTypes = {
+      classes: PropTypes.string,
+      buttonOk: PropTypes.string,
+      buttonCancel: PropTypes.string,
+      locale: localePropType,
+      value: PropTypes.string,
+      startDateYear: PropTypes.number,
+      startDateMonth: PropTypes.number,
+      startDateDay: PropTypes.number,
+      allowedYears: PropTypes.arrayOf(PropTypes.number),
+      allowedYearsRange: PropTypes.string,
+      lowerEndYear: PropTypes.number,
+      higherEndYear: PropTypes.number,
+      startDateMonthTitle: PropTypes.string,
+      startDateYearTitle: PropTypes.string,
+      monthItems: PropTypes.arrayOf(PropTypes.shape({
+        name: PropTypes.string,
+        url: urlPropType,
+        isSelected: PropTypes.bool,
+        value: PropTypes.string,
+      })),
+      yearItems: PropTypes.arrayOf(PropTypes.shape({
+        name: PropTypes.string,
+        url: urlPropType,
+        isSelected: PropTypes.bool,
+        value: PropTypes.string,
+      })),
+    }
+
+    init() {
+      super.init({ styles, template });
+      this.className = `m-datepicker${this.classes ? ` ${this.classes}` : ''}`;
+    }
+
+    range(start, end) {
+      return Array.from({ length: (end - start) }, (v, k) => k + start);
+    }
+
+    connectedCallback() {
+      super.connectedCallback();
+
+      // This method is triggered up to a millions times.... but it's needed here else, no items are available
+      this.props.monthItems = getAllLocaleMonthsArray(this.props.locale).map((item, index) => ({
```

Comment:
@LucaMele 
@jackblackCH 
These values should not change often, right?
So how about memoization or inline caching?

Reply:
@AndyOGo please explain more in detail what you mean or add a code sample. If not possible, @jackblackCH u can ignore this comment

Reply:
@LucaMele 
Well, I asked a question. And gave a suggestion to his inline comment:
> This method is triggered up to a millions times.... but it's needed here else, no items are available

So to clarify my question is:
> These values should not change often, right?

Am I correct with above assumption?

Only if my assumption is correct, this fact of "method is triggered up to a millions times" can be mitigated by caching the functions result or a memoization library.

Reply:
ahhh ok now i understood i didnt get the connection between your comment and the comment on the code :) ok very good suggestions thank you

Reply:
True, it's based on props though. Can we discuss this issue/pattern/problem like in a code review and define some "best practices" or, common pitfalls or whatever? for the moment I could not find a proper solution that's not breaking the whole component.

Reply:
The comment was copied when the method was still in `init`. It is not relevant anymore since the `connectedCallback()` is only called once per component.